ample/ensembler/truncation_util.py

```diff
@@ -32,204 +39,98 @@ def calculate_residues_focussed(var_by_res):
     truncation_levels = percentages
-
-    # print "var_by_res ",var_by_res
     idxs_all = [ x.idx for x in var_by_res ]
     resseq_all = [ x.resSeq for x in var_by_res ]
     variances = [ x.variance for x in var_by_res ]
-
     truncation_residue_idxs = [ sorted(idxs_all[:i + 1]) for i in start_indexes ]
-    # print "truncation_residue_idxs ",truncation_residue_idxs
     truncation_residues = [ sorted(resseq_all[:i + 1]) for i in start_indexes ]
-    # print "truncation_residues ",truncation_residues
-    
     # We take the variance of the most variable residue
     truncation_variances = [ variances[i] for i in start_indexes ] 
-    # print "truncation_variances ",truncation_variances
     
     return truncation_levels, truncation_variances, truncation_residues, truncation_residue_idxs
 
 
-def calculate_residues_percent(var_by_res, percent_interval):
-    """Calculate the list of residues to keep if we are keeping self.percent residues under
-    each truncation bin. The threshold is just the threshold of the most variable residue"""
-    
-    MIN_CHUNK = 3  # We need at least 3 residues for theseus to work
-    length = len(var_by_res)
-    start_idxs = _split_sequence(length, percent_interval, min_chunk=MIN_CHUNK)
+def _calculate_start_indexes_from_fixed_percentages(percent_fixed_intervals, length, all_indexes):
+    "Calculate where in the list of residues each percentage bin starts"
+    def find_closest_index(nresidues_ideal, all_indexes):
+        min_dist = 99999 
+        start = None
+        for idx in all_indexes:
+            rdist = abs(nresidues_ideal - (idx + 1))
+            if  rdist <= min_dist and idx + 1 >= MIN_CHUNK:
+                min_dist = rdist
+                start = idx
+        assert start is not None
+        return start
+
+    indexes = []
+    for interval in percent_fixed_intervals:
+        nresidues_ideal = (length * int(interval)) / 100
+        start = find_closest_index(nresidues_ideal, all_indexes)
+        if start not in indexes:
+            indexes.append(start)
+    return indexes
+
+
+def calculate_residues_by_percent(var_by_res, percent_truncation=None, percent_fixed_intervals=None):
+    """Calculate the list of residues to keep if we are keeping the percentage of residues specified
+    in the list of integers percent_fixed_intervals"""
     
+    if percent_fixed_intervals:
+        if not (isinstance(percent_fixed_intervals, list) and \
+               len(percent_fixed_intervals) > 0 and \
+                min(percent_fixed_intervals) > 0 and \
+                max(percent_fixed_intervals) <= 100):
+            raise RuntimeError("Incorrect percent_fixed_intervals argument: {}".format(percent_fixed_intervals))
+
     # Get list of residue indices sorted by variance - from least to most
     var_by_res.sort(key=lambda x: x.variance, reverse=False)
+    all_indexes = [x.idx for x in var_by_res] # indexes correspond to the number of the residue - 1
```

Comment:
Maybe this is slightly better/more efficient. Would allow us to sort things later on or manipulate the data easier

```python
var_by_res_data = [(x.idx, x.resSeq, x.variance) for x in var_by_res]
all_indexes, all_resseq, variances = zip(*var_by_res_data)
```

Reply:
I actually liked this till I changed the code and then realised it made it a bit harder to understand. As this isn't performance critical I think clarity is to be preferred.